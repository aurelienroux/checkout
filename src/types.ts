export interface product {
  id: string
  name: string
  image: string
  variants: variant[]
  description: string
  defaultVariantId: string
}

export interface variant {
  id: string
  name: string
  price: number
  attributes: {
    name: string
    value: string
  }[]
}

export interface updateAttribute {
  attribute: string
  value: string
}

export interface productToAdd {
  description: string
  image: string
  name: string
  price: number
  productId: string
  variantId: string
}

export interface productInCart {
  description: string
  price: number
  productId: string
  qty: number
  variantId: string
}

export enum attributeEnum {
  SIZE = 'Size',
  FABRIC = 'Fabric'
}

export enum cartActionEnum {
  ADD = 'add',
  REMOVE = 'remove'
}
