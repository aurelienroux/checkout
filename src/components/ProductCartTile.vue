<template>
  <div class="tile">
    <div class="tile__image">
      <img :src="product.image" :alt="product.name" />
    </div>
    <div class="tile__content">
      <h3>{{ product.name }}</h3>
      <p>{{ product.description }}</p>
      <div class="tile__qty">
        <button class="button button--plain" @click="adjustQty(cartActionEnum.REMOVE, product)">-</button>
        <span class="qty">{{ product.qty }}</span>
        <button class="button button--plain" @click="adjustQty(cartActionEnum.ADD, product)">+</button>
        <span class="price">${{ product.price }}</span>
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import Vue from 'vue'
import { productInCart, cartActionEnum } from '@/types'

export default Vue.extend({
  name: 'ProductCartTile',
  props: {
    product: {
      type: Object,
      required: true
    }
  },
  data() {
    return {
      cartActionEnum
    }
  },
  methods: {
    adjustQty(action: string, product: productInCart) {
      this.$store.commit('adjustQty', { action, product })
    }
  }
})
</script>

<style lang="scss" scoped>
.tile {
  display: flex;
  margin-top: 2rem;

  &__image {
    align-items: center;
    background: $image-bg;
    border-radius: 0.5rem;
    display: flex;

    img {
      max-width: 6rem;
    }
  }

  &__content {
    flex: 1;
    padding-left: 1rem;
    text-align: left;

    p {
      font-size: 1.1rem;
      margin: 0;
    }
  }

  &__qty {
    align-items: center;
    display: flex;
    font-weight: 800;
  }

  .qty {
    padding: 0.6rem;
  }

  .price {
    flex: 1;
    text-align: right;
  }
}
</style>
