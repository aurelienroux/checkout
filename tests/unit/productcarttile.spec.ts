import { shallowMount } from '@vue/test-utils'

import ProductCartTile from '@/components/ProductCartTile.vue'
import { cartActionEnum } from '@/types'
import { productMock } from './__mocks__'

const adjustQty = jest.fn()

const wrapper = shallowMount(ProductCartTile, {
  propsData: {
    product: productMock
  },
  methods: {
    adjustQty
  }
})

describe('ProductCartTile component', () => {
  it('renders without crashing', () => {
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find('.tile').exists()).toBe(true)
    expect(wrapper.find('.tile__image').exists()).toBe(true)
    expect(wrapper.find('.tile__content').exists()).toBe(true)
    expect(wrapper.find('h3').exists()).toBe(true)
    expect(wrapper.find('.tile__qty').exists()).toBe(true)
    expect(wrapper.findAll('.button').length).toBe(2)
  })

  it('triggers adjustQty method on click', async () => {
    await wrapper.findAll('.button').at(0).trigger('click')
    expect(adjustQty).toHaveBeenCalledWith(cartActionEnum.REMOVE, productMock)
  })
})
