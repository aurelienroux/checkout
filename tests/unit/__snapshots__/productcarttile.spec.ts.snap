// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCartTile component renders without crashing 1`] = `
<div class="tile">
  <div class="tile__image"><img src="http://image.example" alt="product name"></div>
  <div class="tile__content">
    <h3>product name</h3>
    <p>description</p>
    <div class="tile__qty"><button data-cy="remove-item" class="button button--plain">
        -
      </button> <span data-cy="product-qty" class="qty"></span> <button data-cy="add-item" class="button button--plain">
        +
      </button> <span class="price">$</span></div>
  </div>
</div>
`;
